From f56d73072589a8a147d5afb98c04fd5d9714da7a Mon Sep 17 00:00:00 2001
From: Xi Ruoyao <xry111@xry111.site>
Date: Tue, 1 Sep 2026 01:36:35 +0800
Subject: [PATCH 16/29] Revert "modesetting: avoid memory leak when
 ms_present_check_unflip() returns FALSE"

This reverts commit 309ec5a4a71aa189bcf1eb4263d0d34375051856.

[ Xi: revert for now to resolve conflicts.  Will reapply later. ]
Signed-off-by: Xi Ruoyao <xry111@xry111.site>
---
 hw/xfree86/drivers/modesetting/present.c | 24 +++++++++++-------------
 1 file changed, 11 insertions(+), 13 deletions(-)

diff --git a/hw/xfree86/drivers/modesetting/present.c b/hw/xfree86/drivers/modesetting/present.c
index d80302065..642f7baaf 100644
--- a/hw/xfree86/drivers/modesetting/present.c
+++ b/hw/xfree86/drivers/modesetting/present.c
@@ -409,24 +409,22 @@ ms_present_unflip(ScreenPtr screen, uint64_t event_id)
     PixmapPtr pixmap = screen->GetScreenPixmap(screen);
     xf86CrtcConfigPtr config = XF86_CRTC_CONFIG_PTR(scrn);
     int i;
+    struct ms_present_vblank_event *event;
 
     ms_present_set_screen_vrr(scrn, FALSE);
 
-    if (ms_present_check_unflip(NULL, screen->root, pixmap, TRUE, NULL)) {
-        struct ms_present_vblank_event *event;
-
-        event = calloc(1, sizeof(struct ms_present_vblank_event));
-        if (!event)
-            return;
+    event = calloc(1, sizeof(struct ms_present_vblank_event));
+    if (!event)
+        return;
 
-        event->event_id = event_id;
-        event->unflip = TRUE;
+    event->event_id = event_id;
+    event->unflip = TRUE;
 
-        if (ms_do_pageflip(screen, pixmap, event, -1, FALSE,
-                           ms_present_flip_handler, ms_present_flip_abort,
-                           "Present-unflip")) {
-            return;
-        }
+    if (ms_present_check_unflip(NULL, screen->root, pixmap, TRUE, NULL) &&
+        ms_do_pageflip(screen, pixmap, event, -1, FALSE,
+                       ms_present_flip_handler, ms_present_flip_abort,
+                       "Present-unflip")) {
+        return;
     }
 
     for (i = 0; i < config->num_crtc; i++) {
-- 
2.55.0

